### PR TITLE
Render WebmentionJSTag with a dedicated class

### LINKED ISSUE #99: Optimize JS tag by generating the `template`markup once?
@ashmaroli As you know a lot more about the innards of Jekyll than I do, you can probably shed some light here.

Right now the `webmention_js` tag [assembles template contents each time it is rendered](https://github.com/aarongustafson/jekyll-webmention_io/blob/master/lib/jekyll/tags/webmentions_js.rb#L37-L44). The template names are referenced in the `WebmentionIO` object and then rendered by calling [`Jekyll::WebmentionIO.get_template_contents()`](https://github.com/aarongustafson/jekyll-webmention_io/blob/master/lib/jekyll/webmention_io.rb#L238-L247) which is a disk read every time. Would it make more sense (since the rendering is the same each time the tag is rendered) to temporarily—for the lifetime of `build`—to cache the result of the tag output the first time the tag is rendered and then respond with that if it exists?

Additionally, since other code calls this too, would it make sense to cache each template file’s contents when `Jekyll::WebmentionIO.get_template_contents()` is called?

What would be the best/most Jekyll way to do that?

### PR DESCRIPTION
**Pull request by ashmaroli**
Resolves #99 

The rendered output of `WebmentionJSTag` is the same irrespective of given `context`. So the content can be rendered once, cached elsewhere and then retrieved from the cache for each call to the tag's `:render` method.

(This PR also involves a micro-optimization by caching array of supported_templates in the main module itself, to avoid unnecessary allocations)

### CODE REVIEW

File: lib/jekyll/webmention_io.rb [hunk 1]
```diff
@@ -235,15 +234,36 @@ def self.webmention(source, target, endpoint)
...
-      log "info", "Template file: #{template_file}"
-      File.read(template_file)
+      template_file = template_file(template)
+      @template_content_cache[template_file] ||= begin
+        log "info", "Template file: #{template_file}"
+        File.read(template_file)
+      end
+    end
+
+    def self.gathered_templates
```
Comment (aarongustafson) [inline lib/jekyll/webmention_io.rb]: What do you think about renaming this `html_templates`?

File: lib/jekyll/webmention_io/js_handler.rb [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+      end
+
+      def render
+        if disabled?
+          Jekyll::WebmentionIO.log "info",
+            "JavaScript output is disabled, so the {% webmentions_js %} tag is being skipped"
+          return ""
+        end
+
+        js_file = deploy? ? "<script src=\"#{resource_url}\" async></script>" : ""
```
Comment (aarongustafson) [inline lib/jekyll/webmention_io/js_handler.rb]: Should that be `{@resource_url}`?

**Comment by aarongustafson** [comment]
I like this approach. Thank you!

I added two questions/comments.

PS - I had not seen the ||= assignment pattern before. Using that with return is pretty clever!

File: lib/jekyll/webmention_io/js_handler.rb (hunk 2 above)
Comment (ashmaroli) [reply lib/jekyll/webmention_io/js_handler.rb]: I have an `attr_reader` defined for `resource_url`.
So `#{resource_url} == #{@resource_url}`. It therefore boils down to what's more readable...
(Performance impact is negligible)

**Comment by ashmaroli** [comment]
> I had not seen the ||= assignment pattern before. Using that with return is pretty clever!

:slightly_smiling_face: You're welcome.
Jekyll Core has `||=` used in various places :wink:

File: lib/jekyll/webmention_io/js_handler.rb (hunk 2 above)
Comment (aarongustafson) [reply lib/jekyll/webmention_io/js_handler.rb]: Ok, cool.

**Comment by aarongustafson** [comment]
This had a noticeable speed improvement. Thank you for making it happen!

**Comment by ashmaroli** [comment]
> This had a noticeable speed improvement.

There was none at my end though.. (probably coz I'm on Windows....)

**Pull request merged**